libyul/AsmAnalysis.cpp

```diff
@@ -318,33 +331,39 @@ vector<YulString> AsmAnalyzer::operator()(FunctionCall const& _funCall)
 	for (size_t i = _funCall.arguments.size(); i > 0; i--)
 	{
 		Expression const& arg = _funCall.arguments[i - 1];
-		bool isLiteralArgument = needsLiteralArguments && (*needsLiteralArguments)[i - 1];
-		bool isStringLiteral = holds_alternative<Literal>(arg) && get<Literal>(arg).kind == LiteralKind::String;
-
-		if (isLiteralArgument && isStringLiteral)
-			argTypes.emplace_back(expectUnlimitedStringLiteral(get<Literal>(arg)));
-		else
-			argTypes.emplace_back(expectExpression(arg));
-
-		if (isLiteralArgument)
+		if (auto literalArgumentKind = literalArguments ? literalArguments->at(i - 1) : std::nullopt)
 		{
 			if (!holds_alternative<Literal>(arg))
 				m_errorReporter.typeError(
 					9114_error,
 					_funCall.functionName.location,
 					"Function expects direct literals as arguments."
 				);
-			else if (
-				_funCall.functionName.name.str() == "datasize" ||
-				_funCall.functionName.name.str() == "dataoffset"
-			)
-				if (!m_dataNames.count(std::get<Literal>(arg).value))
-					m_errorReporter.typeError(
-						3517_error,
-						_funCall.functionName.location,
-						"Unknown data object \"" + std::get<Literal>(arg).value.str() + "\"."
-					);
+			else if (*literalArgumentKind != get<Literal>(arg).kind)
+				m_errorReporter.typeError(
+					5859_error,
+					get<Literal>(arg).location,
+					"Function expects " + to_string(*literalArgumentKind) + " literal."
+				);
+			else if (*literalArgumentKind == LiteralKind::String)
+			{
+				if (
+					_funCall.functionName.name.str() == "datasize" ||
+					_funCall.functionName.name.str() == "dataoffset"
+					)
+					if (!m_dataNames.count(get<Literal>(arg).value))
+						m_errorReporter.typeError(
+							3517_error,
+							get<Literal>(arg).location,
+							"Unknown data object \"" + std::get<Literal>(arg).value.str() + "\"."
+						);
+				argTypes.emplace_back(expectUnlimitedStringLiteral(get<Literal>(arg)));
+			}
+			else
+				argTypes.emplace_back(expectExpression(arg));
```

Comment:
Maybe a `continue` in the unlimited string literal case would be better than indented repetition of this?